### ISSUE TITLE
bmv2 extracts a wrong key value

### ISSUE THREAD

**Issue by gareging** [p4lang/behavioral-model#685]
Hello,

I have this header and its parser:

   header trill_t {
        bit<2> v;
        bit<2> reserved;
        bit<1> m;
        bit<5> oplength;
        bit<6> hopcount;
        bit<16> egrb;
        bit<16> igrb;
     }

    state parse_ethernet {
        packet.extract(hdr.ethernet);
        transition select(hdr.ethernet.etherType) {
            0x8912: parse_trill;
            0x0800: parse_ipv4;
            default: accept;
        }
    }

     state parse_trill{
         packet.extract(hdr.trill);
         transition select(hdr.trill.oplength){                                                   
             5w0x0: parse_inner_ethernet;
             default: accept;
         }
     }

**Observed behaviour:**
    Sending a packet.
    11:41:36.683793 bc:87:25:14:1b:ac > 1e:70:34:93:d2:5e, Unknown Ethertype (0x8912), length 62:
     0x0000:  481a 1b74 71a6 e1f5 4b18 1dcb e6af d349 H..tq...K......I
     0x0010:  01e8 0800 4700 001c 0001 0000 4000 142a ....G.......@..*
     0x0020:  54ba c408 25aa bf13 0005 3232 3500 0000 T...%.....225...

Bit #5 in trill's header is set to 1. Oplength (the next 5 bits) set to 0.
Key value is 0x20, which is 32 => requires 6 bits instead of 5.
However, in the code -  **transition select(hdr.trill.oplength)** - we use a 5-bit field hdr.trill.oplength.

[11:41:36.683] [bmv2] [D] [thread 30611] [1.0] [cxt 0] Parser state 'parse_trill': key is 20.

**Expected behaviour**
Key value should be equal to Oplength, i.e. 0.

Please see attached the full p4 code, and a pcap file with two packets: the first one, with the 5th bit set to 0, and the second one as described above. In addition, I attached the switch's log.
[logs.zip](https://github.com/p4lang/behavioral-model/files/2576882/logs.zip)

**Comment by jafingerhut** [comment]
Thanks for the report.

I don't know why this happens yet, but it does look like a bug in behavioral-model code, even the latest version as of 2018-Nov-13 and also latest p4c compiler as of that date.  My best guess would an off-by-one error in some bit-shifting/masking logic in behavioral-model.

I looked at the BMv2 JSON file for the definition of the trill header, and it looks correct.

I added a 'debug table' like shown below to be the first thing that is apply'd during ingress, and when running simple_switch with `--log-console` it is also showing the value of the oplength field as `20`, not `00` as I think it should be.
```
    table debug_table {
        key = {
            hdr.trill.v : exact;
            hdr.trill.reserved : exact;
            hdr.trill.m : exact;
            hdr.trill.oplength : exact;
            hdr.trill.hopcount : exact;
            hdr.trill.egrb : exact;
            hdr.trill.igrb : exact;
        }
        actions = { NoAction; }
        const default_action = NoAction;
    }
```
Below is a Scapy command that can be used to create an input packet that exercises this behavior, and another to send it to interface veth2:
```
p1=Ether(type=0x8912) / Raw('\x48\x1a\x1b\x74\x71\xa6')
sendp(p1, iface='veth2')
```
